### PR TITLE
Support for Python 3.9

### PR DESCRIPTION
**Pull request by crusaderky**
- Supersedes #7166 
- Depends on https://github.com/dask/distributed/pull/4460

**Note:** numba support for Python 3.9 is already available in numba-0.53rc2 on the numba channel; however it contains a regression that causes our tests to fail: https://github.com/numba/numba/issues/6773

### CODE REVIEW

**Comment by crusaderky** [comment]
I'm getting issues on parallel pytest on MacOS.
@jrbourbeau or anybody else with a macbook at hand: could you please run this and report the output?

```bash
conda create -y -n demo_hash_py39 -c conda-forge python=3.9
conda activate demo_hash_py39

python --version
python -c 'print(hash("Hello world"))'
python -c 'print(hash("Hello world"))'
PYTHONHASHSEED=42 python -c 'print(hash("Hello world"))'
PYTHONHASHSEED=42 python -c 'print(hash("Hello world"))'

conda deactivate
conda env remove -n demo_hash_py39
```

File: continuous_integration/environment-3.6.yaml [hunk 1]
```diff
@@ -31,7 +31,7 @@ dependencies:
   - boto3
   - botocore
   - aiohttp
-  - bokeh
+  - bokeh<2.3
```
Comment (crusaderky) [inline continuous_integration/environment-3.6.yaml:34]: https://github.com/dask/dask/issues/7292

Comment (jrbourbeau) [reply continuous_integration/environment-3.6.yaml:34]: We should be able to remove this version constraint now that https://github.com/dask/dask/pull/7297 is in

File: continuous_integration/environment-3.8.yaml [hunk 2]
```diff
@@ -5,8 +5,8 @@ channels:
 dependencies:
   # required dependencies
   - python=3.8
-  - numpy=1.20
-  - pandas=1.2
+  - numpy
+  - pandas
```
Comment (jrbourbeau) [inline continuous_integration/environment-3.8.yaml:9]: We tend to use the different Python version builds to also test against different versions of NumPy and pandas. Could you pin NumPy and pandas for this build to be something between the Python 3.7 and 3.9 builds?

**Review by jrbourbeau — COMMENTED** [review]
This is great to see, thanks for your work on this (and the corresponding PR in `distributed`).

File: continuous_integration/environment-3.8.yaml (hunk 2 above)
Comment (crusaderky) [reply continuous_integration/environment-3.8.yaml:9]: I'm planning to shift all versions in the PR where I drop python 3.6

**Comment by crusaderky** [comment]
The current situation is that Python 3.9 on Windows is unstable. I had 

- a C-level stack overflow on an unspecified test
- an interpreter crash on a trivial test that should trigger a RecursionError

the two things are likely related. I'm going to try and reproduce on my local windows machine.

Everything else is good to go.

File: continuous_integration/environment-3.6.yaml (hunk 1 above)
Comment (crusaderky) [reply continuous_integration/environment-3.6.yaml]: done

**Comment by crusaderky** [comment]
The issue in Windows was caused by ipython, which causes the interpreter to crash instead of raising RecursionError: https://github.com/ipython/ipython/issues/12197

**Comment by crusaderky** [comment]
@jrbourbeau @mrocklin Both PRs are ready for review and merge.
The latest round of tests will fail as I just reverted to distributed@master.
Good tests visible as of the previous commit https://github.com/dask/dask/runs/2013590328

**Review by jrbourbeau — APPROVED** [review]
Thanks @crusaderky! This looks great

**Pull request merged**